peps/pep-0679.rst

```diff
@@ -1,137 +1,216 @@
+
+    >>> print(ast.dump(ast.parse('assert(True, "Error message")'), indent=4))
+    Module(
+        body=[
+            Assert(
+                test=Constant(value=True),
+                msg=Constant(value='Error message'),
+                paren_syntax=1)])
+
+
+Implementing in the compiler
+----------------------------
+
+The new syntax can be implemented in the compiler by special casing tuples
+of length two. This however, will have the side-effect of not modifying the
+AST whatsoever during the transition period while the :exc:`SyntaxWarning`
+is being emitted.
+
+Once the :exc:`SyntaxWarning` is removed, the implementation
+can be moved to the parser level, where the parenthesized form would be
+parsed directly into the same AST structure as ``assert expression, message``.
+This approach is more backwards-compatible, as the many tools that deal with
+ASTs will have more time to adapt.
 
 
 Backwards Compatibility
 =======================
 
-The change is not technically backwards compatible, as parsing ``assert (x,y)``
-is currently interpreted as an assert statement with a 2-tuple as the subject,
-while after this change it will be interpreted as ``assert x,y``.
+The change is not technically backwards compatible. Whether implemented initially
+in the parser or the compiler, ``assert (x,y)``,
+which is currently interpreted as an assert statement with a 2-tuple as the
+subject and is always truth-y, will be interpreted as ``assert x,y``.
 
 On the other hand, assert statements of this kind always pass, so they are
 effectively not doing anything in user code. The authors of this document think
 that this backwards incompatibility nature is beneficial, as it will highlight
-these cases in user code while before they will have passed unnoticed (assuming that
-these cases still exist because users are ignoring syntax warnings).
-
-Security Implications
-=====================
-
-There are no security implications for this change.
+these cases in user code while before they will have passed unnoticed. This case
+has already raised a :exc:`SyntaxWarning` since Python 3.10, therefore these
+cases are expected to be rare, as they require users to ignore existing
+syntax warnings. The continued raising of a :exc:`!SyntaxWarning` should
+mitigate surprises.
+
+The change will also result in changes to the AST of ``assert (x,y)``,
+which currently is:
+
+.. code-block:: text
+
+    Module(
+        body=[
+            Assert(
+                test=Tuple(
+                    elts=[
+                        Name(id='x', ctx=Load()),
+                        Name(id='y', ctx=Load())],
+                    ctx=Load()))],
+        type_ignores=[])
+
+the final implementation, in Python 3.18, will result in the following AST:
+
+.. code-block:: text
+
+    Module(
+        body=[
+            Assert(
+                test=Name(id='x', ctx=Load()),
+                msg=Name(id='y', ctx=Load()))],
+        type_ignores=[])
```

Comment:
Will the `paren_syntax` argument/attribute of `ast`'s `Assert` be removed in 3.18? That would be another backwards-incompatible change :(